### ISSUE TITLE
Organisations: Link to allow users to review their organisation permissions [3]

### ISSUE THREAD

**Issue by cjwatson** [canonical-ols/build.snapcraft.io#680]
The way the GitHub API works for showing organisation repositories is that you ask for the `read:org` scope, and then the OAuth authorisation page asks you which organisations you want the application to be able to see.  If you want to review that list later - and this is quite likely - then you need to go to your [authorised applications](https://github.com/settings/applications) page and configure it there.

The GitHub documentation [recommends](https://developer.github.com/v3/oauth/#directing-users-to-review-their-access-for-an-application) that applications provide a direct link to the GitHub settings page corresponding to their application, so that users can review this sort of thing easily.  We need to work out where to put that.

For comparison, see the [Travis profile page](https://travis-ci.org/profile), which has a bit in the bottom left saying "Is an organization missing?  Review and add your authorized organizations."

**Comment by matthewpaulthomas** [comment]
Most obviously, this will be an issue when you choose “Add repos” and don’t see the repo you’re looking for. That has two variations:
1\. Your only public repos belong to this organization. So you don’t see any repos at all, so you get the empty state.
2\. You have public repos that are personal, or that belong to organizations for which you’ve authorized Snapcraft. So you see some of the repos you expect, but not all of them.

@cjwatson, what would happen if you add a repo, then on GitHub you (accidentally, perhaps) revoke build.snapcraft.io’s access to that repo’s organization, then you return to the “My repos” page? Would the repo just disappear from the list altogether? Or would/could it appear as a ghost, where you could (for example) click on it to get an explanation directing you to the GitHub settings page?

**Comment by cjwatson** [comment]
I think the only realistic option would be for it to disappear altogether.  We can only know which repositories to display by way of knowing your organisation membership, and if the permission has been revoked then GitHub won't even mention that organisation to us.  The only other available data store that could tell us what repositories a particular user has added would be the snap store, but (a) that requires signing in to SSO and (b) it can only tell us about repositories where the user has registered a name, so I don't think that approach is viable here.

It may be worth considering the Travis-style option of showing you a list of organisations that BSI knows you're a member of, which would be a useful cue.

**Comment by matthewpaulthomas** [comment]
Okay. I don’t think we should link to the settings from the “My repos” page, just for the sake of people who have removed build.snapcraft.io’s access to an organization that they had previously added a repo from. That’s too few people for the clutter to be worth it. And if a repo does go missing for that reason, the first thing they’ll try to do is try to re-add it, and when they don’t see it in the “Add repo” list either, then they’re on the same user journey as someone who hadn’t added the repo yet.

**Comment by cjwatson** [comment]
I'm OK with that.  It just needs to be somewhere; IIRC when I first went through this in a local test the GitHub OAuth authorisation page defaulted to none of the organisations I'm in being checked, so it's very easy to accidentally not enable your organisations on the first time through.

**Comment by matthewpaulthomas** [comment]
Understood. Okay, three more questions:

A. For GitHub organizations that an app like build.snapcraft.io has not been granted access to, does the app know (a) that they exist, (b) that they exist and what their names are, or (c) neither? (I’m guessing the answer is neither.)

B. If the answer is neither, then for (1) the new empty state suggestion, does this text seem reasonable?
“Denied **organization access**? You haven’t given Snapcraft access to any of your organizations.
( Review organization access… )”
or, if the number is non-zero,
“Denied **organization access**? You’ve given Snapcraft access to {number} of your organizations.
( Review organization access… )”

(A problem with listing the organizations is that if there are many of them, the other possible reasons for no repos showing up at all would be shunted down out of view, because the dialog is — or will be — fixed height. But then, for (2) the normally-populated Add dialog, I think we should use a popup for missing organizations, like the “Why?” popup for organization/private repos. That popup _would_ have room to list all the organizations, but it would be a bit silly to list them in one situation and not the other. Maybe we could list up to three in the empty-state case, followed by “and {number} others”. Or maybe we could list all the organizations in both cases, but order the organization access suggestion last, so that it doesn’t hide any of the other suggestions…)

C. If you do hit the button, go off to GitHub, and give build.snapcraft.io access to more organizations than it had before … then what? Is there a hook for build.snapcraft.io to notice immediately that you’ve done this? Or would you have to reload the page? Travis navigates to the permissions page in the same window/tab, taking you out of Travis entirely; that seems like a bad idea, unless you’re going to have to reload the Travis page to see the newly-granted organizations anyway.

**Comment by cjwatson** [comment]
A. Neither (I believe).

B. I'm not sure we can tell the difference between "user is not in any orgs" and "user has denied access to all their orgs", which seems perhaps relevant here.  Also "Denied organization access?" feels slightly jargony to me.  But aside from those things your reasoning seems to make sense.

C. I don't know of a relevant hook, although it would not be necessary to reload the page, merely to cancel the "Add repositories" dialog and re-enter it.  That doesn't feel *too* unreasonable to me.

**Comment by cjwatson** [comment]
Perhaps "Can't see organization repos?" rather than "Denied organization access?"

**Comment by matthewpaulthomas** [comment]
I’ve changed “You haven’t given Snapcraft access to any of your organizations” to “You haven’t given Snapcraft access to any organizations”, to address the possibility that you might not have any.

“Denied” is a bit harsh, especially since — as you said above — not-granting is the default, so you might not have explicitly denied anything. And I used “organization access” because that’s the phrase used on the GitHub auth page itself, but I guess it’s enough to use that phrase just in the button label. Changed to “Missing an **organization**?”.

I’ll spec that once you’ve clicked “Review organization access…”, an “OK, I’ve changed it” button should appear alongside to trigger a reload.

**Comment by matthewpaulthomas** [comment]
[Spec updated](https://docs.google.com/document/d/1koTwI3cRr2S2vj5zHKKbR7aWxLRfShDVq2gHQ29cDyU/edit#heading=h.43owh6tjv12q). The downside is that this will involve a bit more front-end work than might have been expected. (Sorry!) The upside is that the overall UI will actually get simpler.

The reason is that up till now, if you could see some repos but not the one you want, there were only two possible reasons: it’s a private repo, or it’s an organization repo. And for both, the action was the same: sign up to be e-mailed when we’re ready. So those could be handled in a small “Why?” popup, containing one text field and one button, on top of the dialog.

But now, there will be three possible reasons: it’s a private repo, you haven’t granted access to the organization, or you don’t have admin permission. Those reasons all have their own text, and two of them have their own form controls. That’s too much stuff to hold in a small popup: it would rival its parent dialog in size. Alternatively, having two separate links in the dialog, to open two separate popups about three reasons that repos might be missing, would be cluttersome and error-prone.

It always seemed a bit awkward to me that I’d designed the same sign-up form to appear in two different ways. This case, plus the you’re-not-the-admin case, kinda forces my hand: we should present these three reasons for missing repos in the same way, empty state or not.

So, to summarize the changes:
- The empty state gains a new suggestion: “Missing an **organization**? You’ve given Snapcraft access to {number} organizations: {comma-separated list, ending with ellipsis if necessary}”, with buttons “Review organization access…” and “OK, I’ve added it”.
- “Private and organization repos not shown. (**Why?**)” at the top is replaced by “{number} selected of {number} repos” at the bottom (**Any missing?**)”. This puts the UI for dealing with repos we aren’t showing right next to the number of repos we _are_ showing.
- There is no more separate sign-up popup. Instead, if you click “Any missing?”, the list of repos should be replaced by the empty state, just omitting the empty-state intro about there being no repos at all.

**Comment by matthewpaulthomas** [comment]
I’ve expanded the spec to clarify the three states for the Add dialog: empty state, populated state, and the state if you choose “Any missing?”.

**Comment by bartaz** [comment]
Do we have visual style for that? Or is spec all we have for now?

Spec seem to assume "Add repos" is has fixed height scrollable view, which is not the case in current implementation and as far as I understand it was meant to be discussed and designed as part of #659

So current proposal of "Anything missing" popover panel seems to be blocked on the changes to "Add repos" view - unless some other visual style can be used until "Add repos" layout and scrolling is updated in #659.

**Comment by matthewpaulthomas** [comment]
We don’t have a visual designer this week, sorry.

Yes, the wireframe does assume that the “Any missing?” link at the bottom will always be visible, which would require that the dialog be shorter than the viewport, because that’s necessary for other things anyway (#564).

It would be possible to fix this issue without doing that as well, by just changing “Private and organization repos not shown. (Why?)” to “(Any repos missing from this list?)” immediately below the dialog title, with the suggestions panel opening downward from that instead of upward. That would be going off-piste from the design, but then the current implementation is already off-piste from the design. It’s up to you.

**Comment by bartaz** [comment]
@matthewpaulthomas ok, thanks for update. Re-using existing popover seems a good compromise (especially that it serves similar purpose now).

The new "Anything missing?" popover has a lot more content then just subscribe form that is in popover now, so it may not look great in 'tooltip' style, but to avoid blocking it it's probably best to try this approach and refine later.
